### ISSUE TITLE
~30s pause when running tests

### ISSUE THREAD

**Issue by charleskorn** [JetBrains/spek#782]
I'm not sure if this is something in my setup or an issue in Spek, but when I run the tests in https://github.com/charleskorn/batect with `./gradlew app:test`, Gradle seems to pause for nearly 30 seconds at `:app:test > 0 tests completed` before then starting to run my tests.

Do you have any ideas on what the cause of this might be? 

Running the tests with a profiler suggests that most of this time is spent calling [`resolveSpec`](https://github.com/spekframework/spek/blob/2.x/spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/SpekRuntime.kt#L24) for each test class.

**Comment by raniejade** [comment]
That usually means you have some long running logic during the discovery phase. i.e

```kotlin
describe("something") {
  runLongRunningThing()

  it("some test") {
    ...
  }
}
```

**Comment by raniejade** [comment]
Probably this is the culprit? https://github.com/charleskorn/batect/blob/master/app/src/journeyTest/kotlin/batect/journeytests/VersionInfoJourneyTest.kt#L32

**Comment by charleskorn** [comment]
Hey @raniejade, thanks for the quick response! `./gradlew :app:test` only runs the tests in https://github.com/charleskorn/batect/tree/master/app/src/unitTest (the journey test you linked to there runs quite quickly).

**Comment by charleskorn** [comment]
Looking at this more closely, it looks like the time is spent loading classes and waiting for Mockito to generate mocks. Would you be interested in a PR that runs the spec resolution in parallel? (I'm imagining something like replacing the `map` call [here](https://github.com/spekframework/spek/blob/2.x/spek-runtime/src/commonMain/kotlin/org/spekframework/spek2/runtime/SpekRuntime.kt#L22) with a call to the `pmap` described at https://jivimberg.io/blog/2018/05/04/parallel-map-in-kotlin/)

**Comment by raniejade** [comment]
Is mockito thread-safe though?

**Comment by charleskorn** [comment]
Yes: https://github.com/mockito/mockito/wiki/FAQ#is-mockito-thread-safe

**Comment by raniejade** [comment]
Nice, I'm currently working on integrating coroutines internally - let's hold this for now until I'm done with it.

**Comment by charleskorn** [comment]
Hey @raniejade, are you still working on integrating coroutines? I'm still keen to have a crack at this

**Comment by raniejade** [comment]
Hey @charleskorn! Yes, you find it at #835. Looks like everything is working, just need to do some testing since it is quite a big change.

**Comment by charleskorn** [comment]
OK great, I'll keep an eye on that PR then.

**Comment by raniejade** [comment]
@charleskorn fyi, #835 is now merged!

**Comment by arturbosch** [comment]
Hello @raniejade ,

we recently splitted some of our code to new modules to reduce incremental build and test time.
However we noticed some increase in test discovery since then.
It gets even worse when using `--parallel`. Modules with ~10 tests take 20 seconds to be discovered (https://github.com/detekt/detekt/issues/2902).
In some modules (rules) we have some heavy setup code but in others the tests are plain simple and still hang for many seconds before starting the work.

I've migrated two simple modules to junit jupiter and the test discovery improvement is huge - https://github.com/detekt/detekt/issues/2902#issuecomment-666259022!

We rely heavily on Spek with more than 1700 tests. Do you have any idea how we can improve the startup time? Thanks!

**Comment by raniejade** [comment]
@arturbosch Thanks for the report! I suspect this line https://github.com/spekframework/spek/blob/60c14fedfb85bfad2e2405406309385cc912f1b8/spek-runtime/src/jvmMain/kotlin/org/spekframework/spek2/runtime/JvmDiscoveryContextFactory.kt#L48 is causing the whole classpath to be scanned. I'll try making the scan asynchronous https://github.com/spekframework/spek/blob/60c14fedfb85bfad2e2405406309385cc912f1b8/spek-runtime/src/jvmMain/kotlin/org/spekframework/spek2/runtime/JvmDiscoveryContextFactory.kt#L52 - hopefully it will help. I will add this in my todo list this weekend.

**Comment by arturbosch** [comment]
Thanks for the quick reply and the ongoing effort! It will help a lot.
Would it be possible to restrict the scanning to some user defined packages?
This could potentially also improve the performance.

**Comment by raniejade** [comment]
> Would it be possible to restrict the scanning to some user defined packages?

I can look at that, but I don't think Gradle's implementation of JUnit 5 support is able to pass that info: https://github.com/gradle/gradle/blob/a9f9b343a3ebfa34534d8245908a1f5702271ac9/subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestClassProcessor.java#L117

From what I can see, if you set a test filter in gradle - gradle itself will do the filtering and passed in classes (I'm not sure of the selection criteria here) to JUnit Platform launcher.
